### ISSUE TITLE
How to build locally with a new patch

### ISSUE THREAD

**Issue by jmckitrick** [daviderestivo/homebrew-emacs-head#178]
I'm experimenting with a patch for a bug I found. I built it a few weeks ago with this command:

```
HOMEBREW_NO_INSTALL_FROM_API=1 brew install --build-from-source --verbose --debug emacs-head@30 --with-cocoa --with-no-frame-refocus --with-modern-icon-emacs-icon4 --with-native-comp
```

And in the terminal output, I saw my patch being applied.

But when I tried to build again today, I no longer see the patch being applied. It's in the source code like this:

```
resource "0012-Hacked.patch" do
   url EmacsHeadAT30.get_resource_url("patches/0012-Hacked.patch")
   sha256 "4e94ed2da2bfe4d88d506eee4a1cd200b4dcc294d8ca0dbe9c769bf71bdd4b2d"
end
```

and of course I have a patch file at `patches/0012-Hacked.patch` and as I said, I saw in the output that it was applied before, a few weeks ago.

Here's what I see now:
```
==> Patching
cp -p /Users/jcm/Library/Caches/Homebrew/downloads/9aa035e80d3ab084120a4788c3ebbe29aab330a221dcbc92b8017a51c2ea5aa4--0001-No-frame-refocus-cocoa.patch /private/tmp/emacs-headA30--patch-20230805-12243-niyax3/0001-No-frame-refocus-cocoa.patch
==> Applying 0001-No-frame-refocus-cocoa.patch
patch -g 0 -f -p1 -i /private/tmp/emacs-headA30--patch-20230805-12243-niyax3/0001-No-frame-refocus-cocoa.patch
patching file 'src/frame.c'
cp -p /Users/jcm/Library/Caches/Homebrew/downloads/aedf6bcffeb3a805dd1feae35654986fa3062c6d42dd2ccc6e9f105bf60e1db1--0005-System-appearance.patch /private/tmp/emacs-headA30--patch-20230805-12243-tfxdnf/0005-System-appearance.patch
==> Applying 0005-System-appearance.patch
patch -g 0 -f -p1 -i /private/tmp/emacs-headA30--patch-20230805-12243-tfxdnf/0005-System-appearance.patch
patching file 'src/frame.h'
patching file 'src/nsfns.m'
patching file 'src/nsterm.m'
cp -p /Users/jcm/Library/Caches/Homebrew/downloads/aa727383a475fe50cd4aa2fdcce43dc2ca41fdf5e3518a2171dbc236c2a1b9c7--0008-Fix-window-role.patch /private/tmp/emacs-headA30--patch-20230805-12243-1uo1wjm/0008-Fix-window-role.patch
==> Applying 0008-Fix-window-role.patch
patch -g 0 -f -p1 -i /private/tmp/emacs-headA30--patch-20230805-12243-1uo1wjm/0008-Fix-window-role.patch
patching file 'src/nsterm.m'
==> ./autogen.sh
Checking whether you have the necessary tools...
```

Is there any automatic process during the build that might overwrite or disable application of the patch?

**Comment by daviderestivo** [comment]
@jmckitrick this is today not possible. I guess what would need to be modified is:

```ruby
  def self.get_resource_url(resource)
    if ENV['HOMEBREW_GITHUB_REF']
      branch = ENV['HOMEBREW_GITHUB_REF'].sub("refs/heads/", "")
      "https://raw.githubusercontent.com/daviderestivo/homebrew-emacs-head/" + branch +  "/" + resource
    else
      "https://raw.githubusercontent.com/daviderestivo/homebrew-emacs-head/" + "master" + "/" + resource
    end
  end
```

to support local files. I can try to prepare a PR over the weekend.

**Comment by jmckitrick** [comment]
@daviderestivo I thought for sure that first command allows me to build/install from the current folder, right? And that flag means to build locally from source, no?

**Comment by daviderestivo** [comment]
Lemme make a try and come back to you ..

**Comment by daviderestivo** [comment]
So, it took a bit of time but I guess I should have something working now.

If you wish to add an additional patch to emacs-head for your testing purposes please follow the below steps:

1. Clone the repo
```bash
git clone https://github.com/daviderestivo/homebrew-emacs-head.git
cd homebrew-emacs-head
# It will be soon merged into master so this step it's not more required
git checkout dev
```

2. Add your patch file into `patches` folder
3. Edit the Formula you want to update (e.g.: emacs-head@29.rb by adding:
```ruby
# Add 
  resource "myTestPatch" do
    url ResourcesResolver.get_resource_url("patches/myTestPatch.patch")
    sha256 "<SHA_256>"
  end

  patch do
    url ResourcesResolver.get_resource_url("patches/myTestPatch.patch")
    sha256 "<SHA_256>"
  end
```

4. Install the formula
```
brew install Formula/emacs-head@29.rb --with-cocoa
```

**Comment by jmckitrick** [comment]
@daviderestivo Thanks, I'll try it out.

**Comment by jmckitrick** [comment]
I'm getting an error early in the build process:

```
~ ❯❯❯ brew install emacs-head@30 --with-cocoa --with-modern-icon-emacs-icon4 --with-native-comp                                                                                                                ✘ 130
==> Fetching daviderestivo/emacs-head/emacs-head@30
Error: emacs-head@30: Failed to download resource "emacs-head@30--patch"
Failure while executing; `/usr/bin/env /opt/homebrew/Library/Homebrew/shims/shared/curl --disable --cookie /dev/null --globoff --show-error --user-agent Homebrew/4.1.5\ \(Macintosh\;\ arm64\ Mac\ OS\ X\ 13.4.1\)\ curl/7.88.1 --header Accept-Language:\ en --retry 3 --fail --location --silent --head file:////Users/jmckitrick/patches/0005-System-appearance.patch` exited with 37. Here's the output:
curl: (37) Couldn't open file //Users/jmckitrick/patches/0005-System-appearance.patch
```

**Comment by jmckitrick** [comment]
Trying the instructions you provided, I get this:

```
~/p/homebrew-emacs-head ❯❯❯ HOMEBREW_USE_LOCAL_RESOURCES=1 brew install Formula/emacs-head@30 --with-cocoa --with-modern-icon-emacs-icon4 --with-native-comp                                                     ✘ 1
Usage: brew install [options] formula|cask [...]
.
.
.
Error: invalid option: --with-cocoa
```

**Comment by daviderestivo** [comment]
You're missing the file extension: **.rb**:

```
HOMEBREW_USE_LOCAL_RESOURCES=1 brew install Formula/emacs-head@30**.rb** --with-cocoa --with-modern-icon-emacs-icon4 --with-native-comp
```

**Comment by jmckitrick** [comment]
🤦🏻‍♂️

Thanks, lol.

I made more progress this time, but I need to revise the patch for some conflicts, so I'll do that later today after work and report back here.

**Comment by jmckitrick** [comment]
@daviderestivo First of all, thanks for putting the time into this.

I pulled the latest from emacs-mirror and generated a patch by reversing the commit that causes the issue. However, that patch fails to apply cleanly. The first problem is I cannot find the .rej file!

```
==> Applying 0012-Hacked.patch
patching file 'src/fileio.c'
1 out of 9 hunks failed--saving rejects to 'src/fileio.c.rej'
```

I've searched my entire filesystem over and over, and I cannot find `fileio.c.rej` anywhere!
After that, I can try to fix the patch.
